drivers/misc/devmux/devmux.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2023, Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_devmux
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/misc/devmux/devmux.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+
+struct devmux_config {
+	const struct device **devs;
+	const size_t n_devs;
+};
+
+struct devmux_data {
+	struct k_spinlock lock;
+	size_t selected;
+};
+
+/* The number of devmux devices */
+#define N DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)
+
+static const struct device *devmux_devices[N];
+static const struct devmux_config *devmux_configs[N];
+static struct devmux_data *devmux_datas[N];
+
+static bool devmux_device_is_valid(const struct device *dev)
+{
+	for (size_t i = 0; i < N; ++i) {
+		if (dev == devmux_devices[i]) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static size_t devmux_inst_get(const struct device *dev)
+{
+	for (size_t i = 0; i < N; i++) {
+		if (dev == devmux_devices[i]) {
+			return i;
+		}
+	}
+
+	return SIZE_MAX;
+}
+
+const struct devmux_config *devmux_config_get(const struct device *dev)
+{
+	for (size_t i = 0; i < N; i++) {
+		if (dev == devmux_devices[i]) {
+			return devmux_configs[i];
+		}
+	}
+
+	return NULL;
+}
+
+struct devmux_data *devmux_data_get(const struct device *dev)
+{
+	for (size_t i = 0; i < N; i++) {
+		if (dev == devmux_devices[i]) {
+			return devmux_datas[i];
+		}
+	}
+
+	return NULL;
+}
+
+ssize_t z_impl_devmux_select_get(const struct device *dev)
+{
+	ssize_t index;
+	struct devmux_data *const data = devmux_data_get(dev);
+
+	if (!devmux_device_is_valid(dev)) {
+		return -EINVAL;
+	}
+
+	K_SPINLOCK(&data->lock)
+	{
+		index = data->selected;
+	}
+
+	return index;
+}
+
+#ifdef CONFIG_USERSPACE
+ssize_t z_vrfy_devmux_select_get(const struct device *dev)
+{
+	return z_impl_devmux_select_get(dev);
+}
+#include <syscalls/devmux_select_get_mrsh.c>
+#endif
+
+int z_impl_devmux_select_set(struct device *dev, size_t index)
+{
+	struct devmux_data *const data = devmux_data_get(dev);
+	const struct devmux_config *config = devmux_config_get(dev);
+
+	if (!devmux_device_is_valid(dev) || index >= config->n_devs) {
+		return -EINVAL;
+	}
+
+	K_SPINLOCK(&data->lock)
+	{
+		*dev = *config->devs[index];
+		data->selected = index;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_devmux_select_set(struct device *dev, size_t index)
+{
+	return z_impl_devmux_select_set(dev, index);
+}
+#include <syscalls/devmux_select_set_mrsh.c>
+#endif
+
+static int devmux_init(struct device *dev)
+{
+	size_t inst = devmux_inst_get(dev);
+	struct devmux_data *const data = dev->data;
+	const struct devmux_config *config = dev->config;
+
+	devmux_configs[inst] = config;
+	devmux_datas[inst] = data;
+
+	for (size_t i = 0; i < config->n_devs; ++i) {
+		if (!device_is_ready(config->devs[i])) {
+			return -ENODEV;
+		}
+	}
```

Comment:
Hey would it make sense to not fail the devmux init if some of the muxed devices are not ready and instead fail the select? Was thinking about a use case where some of the devices may not be available (for example NACKing an i2c address or failing an id register check), and then you could use this for finding alternate devices and have some devices not ready as a known good condition.

Reply:
Hmm... what about the initial selection? I guess it would still make sense to fail on init with the initial selection, right?

Otherwise, that works for me! What do you think, @gmarull?

Reply:
Oh I see the problem. Yeah not sure what to do about the initial selection. Problem is that if you fail the init then the devmux device is flagged as not ready and you can't use it at all. Maybe it make sense to always initially select the first ready sub-device? (if any, if none are ready just fail the mux device as well)

Reply:
With initial selection, it actually narrows the scope of initialization checks, which I think is a good thing. Technically speaking, only the initially selected device needs to be ready by the time the devmux is initialized.

But to your point, it absolutely does make sense to allow possibly tardy devices on i2c, spi, or other buses to have their readiness checked only before selection.

I think that's a quite reasonable compromise. At some point in the future, maybe it would make sense to even allow for deferred initial selection (maybe via DT property?). I'm in no rush to add that though, as my use case primarily involves the console and shell.